modules/amm-executor/src/main/scala/org/ergoplatform/dex/executor/amm/services/CFMMBacklog.scala

```diff
@@ -0,0 +1,202 @@
+package org.ergoplatform.dex.executor.amm.services
+
+import cats.effect.{Clock, Sync}
+import cats.{FlatMap, Monad}
+import derevo.derive
+import org.ergoplatform.dex.domain.amm.{CFMMOrder, OrderId, WeightedOrder}
+import org.ergoplatform.dex.executor.amm.repositories.CFMMOrders
+import tofu.concurrent.{Atom, MakeAtom}
+import cats.syntax.traverse._
+import cats.syntax.option._
+import org.ergoplatform.dex.executor.amm.config.BacklogConfig
+import tofu.concurrent.MakeAtom._
+import tofu.generate.GenRandom
+import tofu.higherKind.Mid
+import tofu.higherKind.derived.representableK
+import tofu.lift.IsoK
+import tofu.logging.{Logging, Logs}
+import tofu.syntax.logging._
+import tofu.syntax.monadic._
+import tofu.syntax.embed._
+import tofu.syntax.time.now
+
+import scala.collection.mutable
+
+@derive(representableK)
+trait CFMMBacklog[F[_]] {
+
+  /** Put an order to the backlog.
+    */
+  def put(order: CFMMOrder): F[Unit]
+
+  /** Put an order with priceTooHigh or priceTooLow execution err
+    */
+  def suspend(order: CFMMOrder): F[Unit]
+
+  /** Put possibly executed order to the backlog
+    */
+  def checkLater(order: CFMMOrder): F[Unit]
+
+  /** Get candidate order for execution. Blocks until an order is available.
+    */
+  def get: F[Option[CFMMOrder]]
+
+  /** Put an order from the backlog.
+    */
+  def drop(id: OrderId): F[Unit]
+}
+
+object CFMMBacklog {
+
+  def make[I[_]: Sync, F[_]: Sync: BacklogConfig.Has: Clock: GenRandom](implicit
+    logs: Logs[I, F],
+    cfmmOrders: CFMMOrders[F],
+    isoKGI: IsoK[F, I]
+  ): I[CFMMBacklog[F]] = for {
+    implicit0(logging: Logging[F]) <- logs.forService[CFMMBacklog[F]]
+    pendingQueueRef                <- MakeAtom[I, F].of(mutable.PriorityQueue.empty[WeightedOrder])
+    suspendedQueueRef              <- MakeAtom[I, F].of(mutable.PriorityQueue.empty[WeightedOrder])
+    revisitOrders                  <- MakeAtom[I, F].of(List.empty[WeightedOrder])
+    _                              <- isoKGI.to(recoverPendingQueue[F](pendingQueueRef, cfmmOrders))
+  } yield BacklogConfig.access
+    .map(cfg =>
+      new CFMMBacklogTracingMid[F] attach new Live[F](
+        pendingQueueRef,
+        suspendedQueueRef,
+        revisitOrders,
+        cfmmOrders,
+        cfg
+      )
+    )
+    .embed
+
+  final private class Live[F[_]: Monad: Clock: GenRandom](
+    pendingQueueRef: Atom[F, mutable.PriorityQueue[WeightedOrder]],
+    suspendedQueueRef: Atom[F, mutable.PriorityQueue[WeightedOrder]],
+    revisitOrders: Atom[F, List[WeightedOrder]],
+    cfmmOrders: CFMMOrders[F],
+    backlogConfig: BacklogConfig
+  ) extends CFMMBacklog[F] {
+
+    /** Put an order to the backlog.
+      */
+    def put(order: CFMMOrder): F[Unit] =
+      cfmmOrders.put(order) >> pendingQueueRef.update(_ += WeightedOrder.fromOrder(order))
+
+    /** Put an order with priceTooHigh or priceTooLow execution err
+      */
+    def suspend(order: CFMMOrder): F[Unit] =
```

Comment:
If I understand correctly, you use cache on `put` and `drop` methods in order to restore queues on app restore, but why you don't use it in `suspend` method?